File: deck/base/code/native/kotlin/runtime/crypto.kt
import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object crypto {
    private fun digestBytes(algorithm: String, input: ByteArray): ByteArray = MessageDigest.getInstance(algorithm).digest(input)
    fun sha256(input: ByteArray): ByteArray = digestBytes("SHA-256", input)
    fun sha512(input: ByteArray): ByteArray = digestBytes("SHA-512", input)
    fun md5(input: ByteArray): ByteArray = digestBytes("MD5", input)
    private fun mac(algorithm: String, key: ByteArray, data: ByteArray): ByteArray {
        val instance = Mac.getInstance(algorithm)
        instance.init(SecretKeySpec(key, algorithm))
        return instance.doFinal(data)
    }
    fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray = mac("HmacSHA256", key, data)
    fun hmacSha512(key: ByteArray, data: ByteArray): ByteArray = mac("HmacSHA512", key, data)
    fun randomBytes(size: Long): ByteArray {
        val bytes = ByteArray(size.toInt())
        java.security.SecureRandom().nextBytes(bytes)
        return bytes
    }
}

File: deck/base/code/native/swift/runtime/crypto.swift
import Foundation
import CryptoKit

enum crypto {
    static func sha256(_ input: Data) -> Data { return Data(SHA256.hash(data: input)) }
    static func sha512(_ input: Data) -> Data { return Data(SHA512.hash(data: input)) }
    static func md5(_ input: Data) -> Data { return Data(Insecure.MD5.hash(data: input)) }
    static func hmacSha256(_ key: Data, _ data: Data) -> Data {
        let mac = HMAC<SHA256>.authenticationCode(for: data, using: SymmetricKey(data: key))
        return Data(mac)
    }
    static func hmacSha512(_ key: Data, _ data: Data) -> Data {
        let mac = HMAC<SHA512>.authenticationCode(for: data, using: SymmetricKey(data: key))
        return Data(mac)
    }
    static func randomBytes(_ size: Int) -> Data {
        var generator = SystemRandomNumberGenerator()
        var bytes = [UInt8]()
        for _ in 0..<size { bytes.append(UInt8.random(in: UInt8.min...UInt8.max, using: &generator)) }
        return Data(bytes)
    }
}
